R/cdisc_data.R

```diff
@@ -0,0 +1,165 @@
+#' @param join_keys (`JoinKeys`) or a single (`JoinKeySet`)\cr
+#'   (optional) object with datasets column names used for joining.
+#'   If empty then it would be automatically derived basing on intersection of datasets primary keys
+#'
+#' @return a `TealData` object
+#'
+#' @details This function checks if there were keys added to all data sets
+#'
+#' @export
+#'
+#' @examples
+#' library(scda)
+#'
+#' latest_data <- synthetic_cdisc_data("latest")
+#' ADSL <- latest_data$adsl
+#' ADTTE <- latest_data$adtte
+#'
+#' # basic example
+#' cdisc_data(
+#'   cdisc_dataset("ADSL", ADSL),
+#'   cdisc_dataset("ADTTE", ADTTE),
+#'   code = "ADSL <- synthetic_cdisc_data('latest')$adsl
+#'           ADTTE <- synthetic_cdisc_data('latest')$adtte",
+#'   check = TRUE
+#' )
+#'
+#' # Example with keys
+#' cdisc_data(
+#'   cdisc_dataset("ADSL", ADSL,
+#'     keys = c("STUDYID", "USUBJID")
+#'   ),
+#'   cdisc_dataset("ADTTE", ADTTE,
+#'     keys = c("STUDYID", "USUBJID", "PARAMCD"),
+#'     parent = "ADSL"
+#'   ),
+#'   join_keys = join_keys(
+#'     join_key(
+#'       "ADSL",
+#'       "ADTTE",
+#'       c("STUDYID" = "STUDYID", "USUBJID" = "USUBJID")
+#'     )
+#'   ),
+#'   code = "ADSL <- synthetic_cdisc_data('latest')$adsl
+#'           ADTTE <- synthetic_cdisc_data('latest')$adtte",
+#'   check = TRUE
+#' )
+cdisc_data <- function(...,
+                       join_keys = teal.data::join_keys(),
+                       code = "",
+                       check = FALSE) {
+  data_objects <- list(...)
+  checkmate::assert_list(
+    data_objects,
+    types = c("TealDataset", "TealDatasetConnector", "TealDataConnector")
+  )
+  if (inherits(join_keys, "JoinKeySet")) {
+    join_keys <- teal.data::join_keys(join_keys)
+  }
+
+  update_join_keys_to_primary(data_objects, join_keys)
+
+  retrieve_parents <- function(x) {
+    tryCatch(
+      x$get_parent(),
+      error = function(cond) rep(character(0), length(x$get_datanames()))
+    )
+  }
+
+  new_parents_fun <- function(data_objects) {
+    lapply(data_objects, function(x) {
+      if (inherits(x, "TealDataConnector")) {
+        unlist(new_parents_fun(x$get_items()), recursive = FALSE)
+      } else {
+        list(retrieve_parents(x))
+      }
+    })
+  }
+
+  new_parents <- unlist(new_parents_fun(data_objects), recursive = FALSE)
+
+  names(new_parents) <- unlist(lapply(data_objects, function(x) {
+    if (inherits(x, "TealDataConnector")) {
+      lapply(x$get_items(), function(z) z$get_dataname())
+    } else {
+      x$get_datanames()
+    }
+  }))
+
+  if (is_dag(new_parents)) {
+    stop("Cycle detected in a parent and child dataset graph.")
+  }
+  join_keys$set_parents(new_parents)
+  join_keys$update_keys_given_parents()
+
+  x <- TealData$new(..., check = check, join_keys = join_keys)
+
+  if (length(code) > 0 && !identical(code, "")) {
+    x$set_pull_code(code = code)
+  }
+
+  x$check_reproducibility()
+  x$check_metadata()
+  return(x)
+}
```

Comment:
Unfortunately, this code overwrites manually specified keys.

```
library(scda)


latest_data <- synthetic_cdisc_data("latest")
ADSL <- latest_data$adsl
ADTTE <- latest_data$adtte
ADSL <- dplyr::rename(ADSL, SID = STUDYID) 
ADTTE <- dplyr::rename(ADTTE, SID = STUDYID)

# Example with keys
cdisc_data(
  cdisc_dataset("ADSL", ADSL,
    keys = c("STUDYID", "USUBJID")
  ),
  cdisc_dataset("ADTTE", ADTTE,
    keys = c("STUDYID", "USUBJID", "PARAMCD"),
    parent = "ADSL"
  ),
  join_keys = join_keys(
    join_key(
      "ADSL",
      "ADTTE",
      c("SID" = "SID", "USUBJID" = "USUBJID")
    )
  )
)
# Error in x$check_metadata() :
#  The join key specification requires dataset ADSL to contain the following columns: STUDYID, USUBJID
```

I wonder if setting default keys shoud be like this. Or maybe we should create `join_keys()` based on the datasets - this will give us default values of the join_keys in the beginning. Alternatively, you can make an `if` here to distinguish if the keys have been specified or not.

```suggestion
cdisc_data <- function(...,
                       join_keys = teal.data::join_keys(...),
                       code = "",
                       check = FALSE) {
  data_objects <- list(...)
  checkmate::assert_list(
    data_objects,
    types = c("TealDataset", "TealDatasetConnector", "TealDataConnector")
  )
  
  x <- TealData$new(..., check = check, join_keys = join_keys)

  if (length(code) > 0 && !identical(code, "")) {
    x$set_pull_code(code = code)
  }

  x$check_reproducibility()
  x$check_metadata()
  return(x)
}
```

Reply:
Alternatively, cdisc_data could look like this:
```
cdisc_data <- function(...,
                       join_keys = teal.data::join_keys(),
                       code = "",
                       check = FALSE) {
  data_objects <- list(...)
  checkmate::assert_list(
    data_objects,
    types = c("TealDataset", "TealDatasetConnector", "TealDataConnector")
  )
  
  if (empty join keys) {
    join_keys <- initialize_datasets_keys(data_objects)
  }
  
  x <- TealData$new(..., check = check, join_keys = join_keys)
  if (length(code) > 0 && !identical(code, "")) {
    x$set_pull_code(code = code)
  }
  x$check_reproducibility()
  x$check_metadata()
  return(x)
}

```

Reply:
The error returned by the app above makes sense and it accounted for because you are trying to define a `cdisc_dataset` with keys containing `STUDYID` where `STUDYID` is not in the data. Try changing `STUDYID` to `SID` and it will work smoothly.